acurl/src/session.pyx

```diff
@@ -0,0 +1,258 @@
+    ):
+        if not isinstance(method, bytes):
+            raise ValueError("method must be bytes")
+        if not isinstance(url, str):
+            raise ValueError("url must be a string")
+        headers = dict(headers)
+        if cookies is None:
+            cookies = {}
+        if json is not None:
+            if data is not None:
+                raise ValueError("use only one or none of data or json")
+            data = dumps(json).encode()  # FIXME: make fast
+            headers.setdefault("Content-Type", "application/json")
+        elif data is not None:
+            if isinstance(data, str):
+                data = data.encode()
+            else:
+                if not isinstance(data, bytes):
+                    # FIXME: not very duck type
+                    raise ValueError("data must be str or bytes")
+
+        # FIXME: probably need to do some escaping if one of the values has
+        # newlines in it...
+        cdef tuple headers_tuple = tuple(k.encode() + b": " + v.encode() for k, v in headers.items())
+        cdef tuple cookie_tuple = tuple(session_cookie_for_url(url, k, v).format() for k, v in cookies.items())
+
+        cdef _Response response = await self._inner_request(
+            method,
+            url,
+            headers_tuple,
+            cookie_tuple,
+            auth,
+            data,
+            cert,
+        )
+        cdef _Response old_response
+        if self.response_callback:
+            # FIXME: should this be async?
+            self.response_callback(response)
+        if (allow_redirects and (300 <= response.status_code < 400) and response.redirect_url is not None):
+            while (max_redirects > 0):
+                max_redirects -= 1
+                if response.status_code in {301, 302, 303}:
+                    method = b"GET"
+                    data = None
+                old_response = response
+                response = await self._inner_request(
+                    method,
+                    response.redirect_url,
+                    headers_tuple,
+                    cookie_tuple,
+                    auth,
+                    data,
+                    cert,
+                )
+                if self.response_callback:
+                    # FIXME: should this be async?
+                    self.response_callback(response)
+                response._set_prev(old_response)
+                if not ((300 <= response.status_code < 400)
+                        and response.redirect_url is not None):
+                    break
+            else:
+                raise RequestError("Max Redirects")
+        return response
+
+    def set_response_callback(self, callback):
+        self.response_callback = callback
+
+    async def get(self, *args, **kwargs):
+        return await self._outer_request(b"GET", *args, **kwargs)
+
+    async def post(self, *args, **kwargs):
+        return await self._outer_request(b"POST", *args, **kwargs)
```

Comment:
add the other http methods: head, delete, options, put

Reply:
Added all of the http methods suggested above 👍